sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala

```
@@ -0,0 +1,212 @@
+    }
+  }
+
+  override def supportsExternalMetadata(): Boolean = false
+}
+
+object StateDataSource {
+  val PARAM_PATH = "path"
+  val PARAM_BATCH_ID = "batchId"
+  val PARAM_OPERATOR_ID = "operatorId"
+  val PARAM_STORE_NAME = "storeName"
+  val PARAM_JOIN_SIDE = "joinSide"
+
+  object JoinSideValues extends Enumeration {
+    type JoinSideValues = Value
+    val left, right, none = Value
+  }
+
+  case class StateSourceOptions(
+      resolvedCpLocation: String,
+      batchId: Long,
+      operatorId: Int,
+      storeName: String,
+      joinSide: JoinSideValues) {
+    def stateCheckpointLocation: Path = new Path(resolvedCpLocation, "state")
+  }
+
+  object StateSourceOptions {
+    def apply(
+        sparkSession: SparkSession,
+        hadoopConf: Configuration,
+        properties: util.Map[String, String]): StateSourceOptions = {
+      apply(sparkSession, hadoopConf, new CaseInsensitiveStringMap(properties))
+    }
+
+    def apply(
+        sparkSession: SparkSession,
+        hadoopConf: Configuration,
+        options: CaseInsensitiveStringMap): StateSourceOptions = {
+      val checkpointLocation = Option(options.get(PARAM_PATH)).orElse {
+        throw new IllegalArgumentException(s"'$PARAM_PATH' must be specified.")
+      }.get
+
+      val resolvedCpLocation = resolvedCheckpointLocation(hadoopConf, checkpointLocation)
+
+      val batchId = Option(options.get(PARAM_BATCH_ID)).map(_.toLong).orElse {
+        Some(getLastCommittedBatch(sparkSession, resolvedCpLocation))
+      }.get
+
+      if (batchId < 0) {
+        throw new IllegalArgumentException(s"'${PARAM_BATCH_ID} cannot be negative.")
+      }
+
+      val operatorId = Option(options.get(PARAM_OPERATOR_ID)).map(_.toInt)
+        .orElse(Some(0)).get
+
+      if (operatorId < 0) {
+        throw new IllegalArgumentException(s"'${PARAM_OPERATOR_ID} cannot be negative.")
+      }
+
+      val storeName = Option(options.get(PARAM_STORE_NAME))
```

Will this set store name to default if its provided as empty string ?

---

No. Do we want to take care of empty string? I don't expect users to specify this except the case of looking into a single state store for stream-stream join. For others they don't need to specify this at all.

---

Let's just throw an exception for empty string.